### DISCUSSION TITLE
Release Management

### DISCUSSION THREAD

**Discussion by yamelsenih** [adempiere/adempiere#4101] [Polls]
Ésta discusión simplemente la abro con motivo de organizar la toma de decisiones con respecto a las próximas liberaciones de ADempiere, inicio con la versión en español.

Desde la versión [3.9.3](https://github.com/adempiere/adempiere/releases/tag/3.9.3) hasta la versión [3.9.4](https://github.com/adempiere/adempiere/releases/tag/3.9.4) pasaron tres años y un mes.

Entendiendo que ADempiere es una comunidad de software libre y que cada uno de los que colaboramos con ella tenemos nuestras ocupaciones, creo que es muy bueno definir la frecuencia de liberación y mantener lo decidido. De tal manera que no se quede escrito sino que realmente exista una liberación en la fecha prevista.

**Comment by marcalwestf** [comment]
@yamelsenih opened this discussion to establish the release frequency.

From version [3.9.3](https://github.com/adempiere/adempiere/releases/tag/3.9.3) to version [3.9.4](https://github.com/adempiere/adempiere/releases/tag/3.9.4) transcurred three years and one month.

The poll above allows the collaborators to decide which frequency each one is inclined to.

You are free to participate and write down the reason for your selection.

**Comment by marcalwestf** [comment]
**A continuación, la argumentación de @e-Evolution al respecto, que escribió en _Discord,_ y que he resumido y documentado aquí para tener presentes todos los argumentos.**

_Preguntas que debemos hacernos:_
- _¿que deseamos nosotros lograr con las politicas de liberación ?_ 
- _¿que desean los clientes?_ 

_I LIBERACIÓN SEMANAL
¿Considera el modelo de actualizar cada semana las migraciones a la base instalada de los clientes?
Con el modelo actual, sería muy costoso liberar semanalmente; talvez más fácil utilizando docker.
Sería conveniente que @YamelSenih explicara cómo funcionan para él las liberaciones semanales._

_II SOLUCIÓN INTERMEDIA
En una solucion intermedia, las liberaciones que nosotros llamamos "microversiones" (las que no afectan al diccionario o tienen mejoras) se podrian implementar semanalmente._

_Esto presupone facilitar el mecanismo de parcheo y liberacion de las migraciones.
Ese mecanismo deberia ser mas simple que el de instalar una version nueva compeltamente otra vez.
Adempiere cuenta ya con este mecanismo y es colocar todos las clases parchadas en un archivo llamado patches.jar._

**Comment by marcalwestf** [comment]
Podríamos acordar tener diferentes tipos de liberaciones: 
- una para corregir _errores_ que **no requieran**  cambios en el Diccionario de Aplicación
- una para corregir _errores_ que **requieran**  cambios en el Diccionario de Aplicación
- otra para _mejoras funcionales_  que **no requieran**  cambios en el Diccionario de Aplicación
- y finalmente, una que incluya 
  - _cambios funcionales_ que **involucren o no** cambios en el Diccionario de Aplicación
  - _errores_ que **requieran o no** cambios en el Diccionario de Aplicación

Para comenzar, podríamos decidir por una política de liberación para cada tipo de liberación, ponerla a prueba, y de acuerdo con las experiencias que tengamos, redefinimos la política de liberación para que se acople a nuestra realidad.

**Comment by piracio** [comment]
Tengo entendido que AD tiene varios métodos para actualizar componentes, el
principal motivo del tiempo entre las liberaciones es la meta que se quiere
para decir que un software está listo.
En lo personal por si no lo saben, se contactó a Victor en Marzo 2020 para
arreglar problemas de memory leak, actualización de librerías y mejoras de
seguridad para evitar exploits que tienen versiones anteriores.
Esto inicialmente era un trabajo de unos meses, sin embargo se fueron
incorporando muchas mejoras que eran necesarias desde versiones 3.6 o
anteriores.

Hoy la versión 394, tiene un manejo bueno de la RAM frente a intentos de
acceso, los motores aplicativos fueron cambiados, se pagó 3.6K USD en
herramientas para realizar pruebas de carga.

Esto se realiza para apoyar al proveedor OFB de Chile, se aprendió GIT,
enseñe performance testing seguridad y Victor me enseñó como el modelo AD
funciona a nivel de código, esto tomó tiempo dado que era más fácil
explicar los problemas detectados desde un punto de visto de Ingeniero de
sistemas y DBA Oracle/Postgresql.

AD es un software JAVA, tirarlo dentro de containers no tiene sentido. Debe
existir la posibilidad ( sí ), debe ser la nueva línea, no lo creo, dado
que AD al ser una versión JAVA esta debe tener la posibilidad de correr
RAW. de lo contrario estamos forzando a clientes a tener otra capa además
de JAVA.

Si quieren ser más serios sobre el ciclo de releases, deben definir los
features/bugs que quieren arreglar que va en el próximo release, comenzar a
hacer performance testing cuando esta un feature listo y revisar la
seguridad, sacar un release por qué a pasado mucho tiempo, sobre todo con
todos los problemas de seguridad y performance detectados en el pasado,
sacando versiones semanales es muy malo, si las versiones de AD fueran
perfectas lo puedes hacer diario si quieres.

Antes de conversar este tema, podemos probar como están las versiones
propuestas para saber si resisten carga o no.

Para los que no me conocen, soy Ingeniero Civil Informático de Chile,
trabajé en empresas grandes como DBA Oracle, seguridad y linux. desde el
2000 se trabaja sintonizando máquina Java para los bancos y desde 2015
vengo haciendo trabajos relacionados a performance tuning y Java para
grandes empresas en paralelo apoyo a una empresa llamada OFB desde el 2010
cuando conocí AD 360 donde mis recomendaciones por ABC motivos nunca
llegaron a AD Git y en el 2020 para apoyar a OFB contacte a Victor para
arreglar AD de una buena vez. Performance, seguridad y actualización de
librerías son mis metas personales para este proyecto.

On Wed, 26 Apr 2023 at 09:10, Mario Calderon ***@***.***>
wrote:

> *A continuación, la argumentación de @e-Evolution
> <https://github.com/e-Evolution> al respecto, que escribió en Discord, y
> que he resumido y documentado aquí para tener presentes todos los
> argumentos.*
>
> *Preguntas que debemos hacernos:*
>
>    - *¿que deseamos nosotros lograr con las politicas de liberación ?*
>    - *¿que desean los clientes?*
>
>
>
>
> *I LIBERACIÓN SEMANAL ¿Considera el modelo de actualizar cada semana las
> migraciones a la base instalada de los clientes? Con el modelo actual,
> sería muy costoso liberar semanalmente; talvez más fácil utilizando docker.
> Sería conveniente que @yamelsenih <https://github.com/yamelsenih> explicara
> cómo funcionan para él las liberaciones semanales.*
>
>
> *II SOLUCIÓN INTERMEDIA En una solucion intermedia, las liberaciones que
> nosotros llamamos "microversiones" (las que no afectan al diccionario o
> tienen mejoras) se podrian implementar semanalmente.*
>
>
>
> *Esto presupone facilitar el mecanismo de parcheo y liberacion de las
> migraciones. Ese mecanismo deberia ser mas simple que el de instalar una
> version nueva compeltamente otra vez. Adempiere cuenta ya con este
> mecanismo y es colocar todos las clases parchadas en un archivo llamado
> patches.jar.*
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/adempiere/adempiere/discussions/4101#discussioncomment-5724678>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABQHBRIRWLPNCINIA4QYYF3XDA4WBANCNFSM6AAAAAAXLD5S34>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***
> com>
>


-- 
Saludos,
Horacio Miranda Aguilera.

**Comment by e-Evolution** [comment]
Creo que hemos acordado que tendremos dos tipos de liberaciones macro versiones y micro versiones

# Macro Versiones #
Las macro versiones están enfocadas en agrupar todos las correcciones y  mejoras de un periodo y deberían cumplir con los siguientes elementos:

01.- Un periodo de 6 meses a 1 año
02.- El tag de liberación para estas versiones incrementan un dígito en la tercera posición, por ejemplo 3.9.4 , 3.9.5 , 3.9.6 y los cambios se aplican a las ramas de develop y master
03.- La liberación requiere cambiar el código con fuente la versión en todas partes
04.- La liberación requiere actualizar cualquier librería de java por temas de seguridad y corregir cualquier vulnerabilidad
05.- La liberación requiere generar una base de datos semilla para PostgreSQL y Oracle con lo que esto implica (Aplicar Script de Migración, Generar Constraints de la DB, revisar Vistas y Funciones) 
06.- La liberación requiere realizar pruebas para las ultimas versiones de PostgreSQL y Oracle DB
07.- La liberación requiere un ciclo de pruebas intensivas de toda la funcionalidad.
08.- La liberación requiere regenera el modelo de clases de todas las tablas 
09.- La liberación requiere crear un binario completo  Adempiere_395LTS.tar.gz 
10.- La liberación requiere un release note detallado y publicar los anuncios oficiales 

## Pasos para migrar una macro versión ##
01.- Hacer una nueva instalación de Adempiere con el nuevo binario Adempiere_395LTS.tar.gz en un servidor de prueba
02.- Importar un respaldo de la version de producción 
03.- Hacer configuración del nuevo ambiente RUN_Setup
04.- Ejecutar la migración de utils/RUN_MigrateXML y validar que todo este bien , si presenta problemas debe resolver cualquier proceso durante la migración.
05.- Validar si sus personalizaciones son compatible , si no se requiere migrar el código , las vistas y modificaciones del diccionario de aplicaciones
06.- El cliente al que se realizo la migración tiene un periodo de prueba si esta no presenta inconvenientes la migración es liberada para producción 
07- Ejecutar del paso 1 al 6 en el servidor de producción y sin usuarios activos

# Micro Versiones #
Las micro versiones están enfocadas en agrupar todos las correcciones de un periodo y deberían cumplir con los siguientes elementos:

01.- Un periodo de 1 a 4  meses.
02.- El tag de liberación para estas versiones incrementan un dígito en la cuarta posición, por ejemplo 3.9.5.01 , 3.9.5.02 , 3.9.5.03, 3.9.5.04 y los cambios se aplican a las ramas de develop y master
03.- La liberación requiere actualizar cualquier librería de java por seguridad
04.- La liberación requiere tener todas las pruebas certificadas para cada pull request de corrección de la funcionalidad.
05.- La liberación requiere generar un archivo de patches_3.9.5.01.jar o zkpatches_3.9.5.01.jar

## Pasos para migrar una micro versión ##
01.- En el ambiente de pruebas de la versión actual  de Adempiere copiar el archivo patches_3.9.5.01.jar a Adempiere/lib/patches.jar o zkpatches_3.9.5.01.jar a Adempiere/lib/zkpatches.jar
02.- Ejecutar un despliegue con RUN_silentsetup
03.- Ejecutar la migración de utils/RUN_MigrateXML y validar que todo este bien , si presenta problemas debe resolver cualquier proceso durante la migración Nota : Los script XML solo deben ser para arreglar problemas del diccionario , en ningún caso deben contener mejoras ya que no se regenera el modelo de clases.
04.- Se realizan pruebas por la empresa consultora para validar  que las correcciones están aplicadas.
05.- Ejecutar del paso 1 al 5 en el servidor de producción y sin usuarios activos

Las principales diferencias de una macro a una micro versión el tiempo para prepararlas y la ejecución de las pruebas en la macro versión son intensivas , mientras en la micro version se garantizan por las hechas en PR, por otro lado en el caso de una micro versión basta con actualizar el archivo de patches.jar y hacer un despliegue , las personalizaciones que se tengan no deberían requerir cambios ya que están basado en la misma base de código y solo arreglan problemas.

Así mientras la migración de una macro versión es mas compleja y requiere pruebas intensivas por parte del cliente.  la micro versión solo requiere parchar las clases actuales , hacer un despliegue , las pruebas son garantizadas desde el PR , no se requiere validar las personalizaciones ya que no hay cambios sustanciales y  no se requiere pruebas por parte del cliente.

Por otra parte preparar una version para su liberación me tomar un tiempo estimado de una semana  y se realiza manualmente, las micro versiones podrían ser automáticas  ya que solo se requiere generar el jar de los parches.

Así ya que yo soy el que prepara las liberaciones, solo puedo dedicar tiempo para preparar una macro version una o máximo dos veces al año, hacer el trabajo para una micro versión como si fuera una macro versión no me seria posible ya que el proceso es algo complejo y requiere de un equipo multi disciplinario, es por eso que pensar en un ciclo de liberaciones tiene que considerar todos estos elementos antes de definir los periodos de liberación.

Saludos
Victor

**Comment by EdwinBetanc0urt** [comment]
Saludos a todos, @e-Evolution  los pasos que propones para la liberación de macro versiones y micro versiones están detallados.

Sin embargo veo que quizás lo que mas tiempo podría demorarse, en el caso de las macro versiones seria el paso 03, tengo varias dudas con respecto a ese paso, porque es necesario realizarlo? o si existe alguna manera que pueda automatizar/omitir este paso?, ya que por ejemplo en caso de que una persona agregue una nueva columna a la tabla `C_Invoice` este mismo debería ser responsable de generar las clases modelo `X_C_Invoice`, `I_C_Invoice`.

**Comment by RamiroV13** [comment]
Adjunto acta reunión ADempiere ocurrido el pasado miércoles 26 de abril 2023 donde se habló todo lo relacionado a este post

[PROYECTO ERP ADEMPIERE v1.pdf](https://github.com/adempiere/adempiere/files/11383102/PROYECTO.ERP.ADEMPIERE.v1.pdf)
